### PR TITLE
Handle Continuation scan during Concurrent Scavenger and Metronome

### PR DESCRIPTION
**Pull request by LinHu2016**
For CS and Metronome, due to synchronous nature of their barriers (they read the value of the slot and act on it immediately, for example they may evacuate object for CS and mark object for Metronome), We fully scan java stacks of related continuation during the virtual thread mount.

depends: https://github.com/eclipse/omr/pull/6781

Signed-off-by: Lin Hu <linhu@ca.ibm.com>

### CODE REVIEW

**Comment by pshipton** [comment]
Note we can run system testing in a PR build, i.e. `... test sanity.system <plats> <vers>`

File: runtime/gc_base/ObjectAccessBarrier.hpp [hunk 1]
```diff
@@ -267,6 +269,9 @@ class MM_ObjectAccessBarrier : public MM_BaseVirtual
 	virtual bool postObjectRead(J9VMThread *vmThread, J9Object *srcObject, fj9object_t *srcAddress);
 	virtual bool postObjectRead(J9VMThread *vmThread, J9Class *srcClass, J9Object **srcAddress);
 
+	virtual void preMountContinuation(J9VMThread *vmThread, j9object_t contObject) {}
+	virtual void postUnmountContinuation(J9VMThread *vmThread, j9object_t contObject) {}
+	virtual void registerScavenger(MM_Scavenger *scavenger) {}
```
Comment (amicic) [inline runtime/gc_base/ObjectAccessBarrier.hpp:274]: I would not introduce subclass specific method in a base class. This method should be only in StandardAccessBarrier and not even virtual.

Then, of course, pointer to accessBarrier must be upcasted when invoking registerScavenger. 

Prior to that we can assert 2 things are valid, to make sure that upcast is reasonable (none tried to register to wrong type of accessBarrier) :
- isStandardGC() - this is exact guard that is used when we instantiate StandardAccessBarrier
- scavengerEnabled

File: runtime/gc_modron_standard/StandardAccessBarrier.cpp [hunk 2]
```diff
@@ -265,10 +265,12 @@ MM_StandardAccessBarrier::postObjectStoreImpl(J9VMThread *vmThread, J9Object *ds
 {
 	/* If the source object is NULL, there is no need for a write barrier. */
 	if(NULL != srcObject) {
+#if defined(OMR_GC_CONCURRENT_SCAVENGER)
 		if (_extensions->isConcurrentScavengerEnabled() && !_extensions->isScavengerBackOutFlagRaised()) {
 			Assert_MM_false(_extensions->scavenger->isObjectInEvacuateMemory(dstObject));
```
Comment (amicic) [inline runtime/gc_modron_standard/StandardAccessBarrier.cpp:270]: missed to use cached value

**Comment by amicic** [comment]
jenkins test sanity.system aix jdk8

**Comment by amicic** [comment]
jenkins test sanity win jdk19

**Comment by amicic** [comment]
Hm, sanity.system failure, does not seem to be related:

```
TST stderr j> 2022/10/24 14:37:55.467 Mon Oct 24 14:37:55 UTC 2022 - main : Threads started, waiting for deadlock to establish
TST stderr j> 2022/10/24 14:37:55.467 Mon Oct 24 14:37:55 UTC 2022 - Deadlock Thread 1 : Acquired second lock
TST stderr j> 2022/10/24 14:37:55.518 Mon Oct 24 14:37:55 UTC 2022 - main : Deadlock established
TST stderr j> 2022/10/24 14:38:00.545 Mon Oct 24 14:38:00 UTC 2022 - Deadlock Thread 1 : Interrupted trying to get second lock
TST stderr j> 2022/10/24 14:38:00.546 Mon Oct 24 14:38:00 UTC 2022 - Deadlock Thread 2 : Interrupted trying to get second lock
TST stderr Exception in thread "main" java.lang.AssertionError: Reported Deadlocked Threads are Incorrect
```

**Comment by pshipton** [comment]
There was a bad change merged https://github.com/eclipse-openj9/openj9/pull/15956

File: test/testng.xml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
+<suite name="Suite">
+  <test name="Test">
+    <classes>
+      <class name=".HandBack"/>
+      <class name=".TestGCPolicyNogc"/>
+    </classes>
+  </test> <!-- Test -->
+</suite> <!-- Suite -->
```
Comment (amicic) [inline test/testng.xml:10]: wrong change. 
expected to touch one of module.xml to allow include from gc_base to gc_modron_standard

File: runtime/gc_base/GCExtensions.cpp [hunk 4]
```diff
@@ -284,3 +285,14 @@ MM_GCExtensions::getContinuationObjectListsExternal(J9VMThread *vmThread)
 {
 	return continuationObjectLists;
 }
+
+
+void
+MM_GCExtensions::registerScavenger(MM_Scavenger *scavenger)
+{
+	MM_GCExtensionsBase::registerScavenger(scavenger);
+	if (isStandardGC()) {
```
Comment (amicic) [inline runtime/gc_base/GCExtensions.cpp:294]: let's actually make this an assert

**Comment by amicic** [comment]
jenkins test sanity aix jdk19

File: runtime/gc_base/module.xml [hunk 5]
```diff
@@ -53,6 +53,7 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 			<include path="j9gcstats"/>
 			<include path="$(OMR_DIR)/gc/stats" type="relativepath"/>
 			<include path="j9gcgluejava"/>
+			<include path="j9modronstandard"/>
```
Comment (amicic) [inline runtime/gc_base/module.xml:56]: need copyright update

Comment (amicic) [reply runtime/gc_base/module.xml:56]: from internal builds I see, we need -I../omr/gc/base/standard too

```
In file included from [01m[K../omr/gc/base/Configuration.hpp:33:0[m[K,
                 from [01m[K../gc_modron_standard/StandardAccessBarrier.hpp:37[m[K,
                 from [01m[KGCExtensions.cpp:40[m[K:
[01m[K../gc_glue_java/ConfigurationDelegate.hpp:34:10:[m[K [01;31m[Kfatal error: [m[KConcurrentGC.hpp: No such file or directory
```

**Comment by amicic** [comment]
The only commit since jenkins tests were launched was touching only a module.xml file that is relevant only for IBM internal builds, so I did not relaunch them, however results got hidden:

aix sanity.functional jdk19 all green:
https://openj9-jenkins.osuosl.org/job/PullRequest-OpenJ9/2907/

xLinux sanity.system jdk8, all green except known, but unrelated failures in JLM:
https://openj9-jenkins.osuosl.org/job/PullRequest-OpenJ9/2906/

**Review by amicic — APPROVED** [review]

**Pull request merged**